cpp/src/groupby/hash/groupby_kernels.cuh

```diff
@@ -0,0 +1,296 @@
+ * that column is skipped.
+ *
+ * @note If a target element is NULL, it is assumed that the value of the NULL
+ * element is the identity value of the aggregation operation being performed.
+ * The aggregation operation is performed between the source element and the
+ * identity value, and the target element's bit is set to indicate it is no
+ * longer NULL.
+ *
+ * @note For COUNT, it is assumed the target element can *never* be NULL. As
+ * such, it is expected the target element's bit is already set.
+ *
+ * @param target Table containing the target row
+ * @param target_index Index of the target row
+ * @param source Table cotaning the source row
+ * @param source_index Index of the source row
+ * @param ops Array of operators to perform between the elements of the
+ * target and source rows
+ *---------------------------------------------------------------------------**/
+template <bool values_have_nulls = true>
+__device__ inline void aggregate_row(device_table const& target,
+                                     gdf_size_type target_index,
+                                     device_table const& source,
+                                     gdf_size_type source_index,
+                                     operators* ops) {
+  using namespace bit_mask;
+  thrust::for_each(
+      thrust::seq, thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(target.num_columns()),
+      [target, target_index, source, source_index, ops](gdf_size_type i) {
+        bit_mask_t* const __restrict__ source_mask{
+            reinterpret_cast<bit_mask_t*>(source.get_column(i)->valid)};
+
+        if (values_have_nulls and nullptr != source_mask and
+            not is_valid(source_mask, source_index)) {
+          return;
+        }
+
+        cudf::type_dispatcher(source.get_column(i)->dtype,
+                              elementwise_aggregator<values_have_nulls>{},
+                              *target.get_column(i), target_index,
+                              *source.get_column(i), source_index, ops[i]);
+      });
+}
+
+template <bool nullable = true>
+struct row_hasher {
+  using result_type = hash_value_type;  // TODO Remove when aggregating
+                                        // map::insert function is removed
+  device_table table;
+  row_hasher(device_table const& t) : table{t} {}
+
+  __device__ auto operator()(gdf_size_type row_index) const {
+    return hash_row<nullable>(table, row_index);
+  }
+};
+
+template <bool skip_rows_with_nulls, bool values_have_nulls, typename Map>
+__global__ void build_aggregation_map(
+    Map* map, device_table input_keys, device_table input_values,
+    device_table output_values, operators* ops,
+    bit_mask::bit_mask_t const* const __restrict__ row_bitmask) {
+  gdf_size_type i = threadIdx.x + blockIdx.x * blockDim.x;
+
+  while (i < input_keys.num_rows()) {
+    if (skip_rows_with_nulls and not bit_mask::is_valid(row_bitmask, i)) {
```

Comment:
Avoid duplicating the `i += blockDim.x * gridDim.x;` by inverting the semantics of this conditional and putting lines 251-254 inside the conditional, and the `i += blockDim.x * gridDim.x` after it.

```suggestion
    if (not skip_rows_with_nulls or bit_mask::is_valid(row_bitmask, i)) {
      auto result = map->insert(thrust::make_pair(i, i));
	
      aggregate_row<values_have_nulls>(output_values, result.first->second,
                                       input_values, i, ops);
    }
    i += blockDim.x * gridDim.x;
```

Reply:
Ah, I was looking for a better way to do this. Thanks.

Reply:
I spent a while trying to figure out the cleanest way to do this, and couldn't find a way to do it without _some_ duplication. This just happened to be the smallest amount of duplication I could come up with.